monai/transforms/croppad/functional.py

```diff
@@ -0,0 +1,131 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+A collection of "functional" transforms for spatial operations
+https://github.com/Project-MONAI/MONAI/wiki/MONAI_Design
+"""
+
+from __future__ import annotations
+
+import numpy as np
+import torch
+from torch.nn.functional import pad as pad_pt
+
+from monai.data.meta_obj import get_track_meta
+from monai.data.meta_tensor import MetaTensor
+from monai.transforms.inverse import TraceableTransform
+from monai.transforms.utils import convert_pad_mode, create_translate
+from monai.utils import TraceKeys, convert_to_dst_type, convert_to_tensor
+
+__all__ = ["pad_nd", "pad_func"]
+
+
+def _np_pad(img: torch.Tensor, pad_width: list[tuple[int, int]], mode: str, **kwargs) -> torch.Tensor:
+    img_np = img.detach().cpu().numpy() if isinstance(img, torch.Tensor) else img
+    mode = convert_pad_mode(dst=img_np, mode=mode).value
+    if mode == "constant" and "value" in kwargs:
+        kwargs["constant_values"] = kwargs.pop("value")
+    out = torch.as_tensor(np.pad(img, pad_width, mode=mode, **kwargs))  # type: ignore
+    if isinstance(img, MetaTensor):
+        out = convert_to_dst_type(out, dst=img)[0]
+    return out
+
+
+def _pt_pad(img: torch.Tensor, pad_width: list[tuple[int, int]], mode: str, **kwargs) -> torch.Tensor:
+    pt_pad_width = [val for sublist in pad_width[1:] for val in sublist[::-1]][::-1]
+    # torch.pad expects `[B, C, H, W, [D]]` shape
+    return pad_pt(img.unsqueeze(0), pt_pad_width, mode=mode, **kwargs).squeeze(0)
+
+
+def pad_nd(img: torch.Tensor, to_pad: list[tuple[int, int]], mode: str, **kwargs):
```

Comment:
it seems this `pad_nd` function is only used within this file, should we make it as private? `_pad_nd`.

Thanks.

Reply:
Hi @Nic-Ma, Wenqi suggested it to be available to the user in this [reply](https://github.com/Project-MONAI/MONAI/pull/6031#discussion_r1115749051).

Reply:
OK, I see.

Thanks.